Design document for generic queue transport in CoreWCF (#736)
CoreWCF/CoreWCF: Documentation/DesignDocs/corewcfqueue-design.md

# Objective 

> The objective of this document is to provide details on how to build a generic queue transport inside CoreWCF so that it can enable various queue binding. This document is based on the discussion outlined here https://github.com/CoreWCF/CoreWCF/issues/164 .

Reviewer: Nit: various queue binding**s**

> # Why are we building

Reviewer: Cleaner section title would be "Problem overview" as current title is a sentence fragment.

> When WCF was build there were not many queing technique and there was no concept of cloud. With multiple cloud providers supporting queing technique and many queing technique widely available, we wanted to build a generic transport, so that various queuing mechanism can be supported in CoreWCF.

Reviewer: Spelling/grammar/phrasing:  
When WCF was built there were not many queuing technologies....

Reviewer: Those same changes should be made through the entire paragraph and not just the one sentence.

# Class Diagram
![CoreWCF Generic Queue](/Documentation/DesignDocs/generic_queue_class_diagram.png?raw=true)

# Class Details
#### IQueueMiddlewareBuilder and QueueMiddlewareBuilder
- IQueueMiddlewareBuilder provides signatures for 2 important methods, ```Use``` and ```Build```. 
```csharp
internal interface IQueueMiddlewareBuilder 
{ 

>    IServiceProvider HandshakeServices { get; set; } 

Reviewer: The name HandshakeServices came from the NetTcp implementation which has a handshake performed on connection. This should be named something different, maybe QueueServices or maybe just Services.

   IDictionary<string, object> Properties { get; } 
   IQueueMiddlewareBuilder Use(Func<QueueMessageDispatcherDelegate, QueueMessageDispatcherDelegate> middleware); 
   QueueMessageDispatcherDelegate Build(); 
} 
```

> - ```Use``` adds middleware delegates to the Queue handshake pipeline 

Reviewer: Drop the word handshake.

> - ```Build``` builds the delegate used by this application to process fetching request hold up in queue and process them.

Reviewer: ...to process requests received from the queue.

#### QueueMiddleWareBuilderExtension

> - QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegrate to be invoked middleware pipeline. Simillar to the way implemented in UseMiddlewareFramingConnectionHandshakeExtensions for NetTCP.

Reviewer: Typo: delegrate->delegate  
..to be involed by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions.

> - QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegate to be invoked by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions..

Reviewer: I just noticed you use triple backticks when inserting inline code fragments. Inline you use a single back tick \`like this\` to produce `like this`.  Can you fix throughout the document. I just noticed we don't define `QueueMessageDispatcherDelegate` anywhere. You should probably add it's definition in the code snippet which contains the `IQueueMiddlewareBuilder` delegate.

```suggestion
The `QueueMiddleWareBuilderExtension` class defines the `UseMiddleware` method which searches the provided `TMiddleware` generic parameter type for an `InvokeAsync` method with a method signature that matches the `QueueMessageDispatcherDelegate` delegate. It expects a constructor to exist which takes as the first parameter an instance of a `QueueMessageDispatcherDelegate` delegate, and any remaining parameters to match the optional additional args. The `QueueMessageDispatcherDelegate` delegate passed to the constructor is the next middleware to call when the middleware doesn't terminate the dispatch chain. This is similar to the NetTcp class `UseMiddlewareFramingConnectionHandshakeExtensions`, and the ASP.NET Core `UseMiddlewareExtensions` class.
```

```csharp

> public class QueueMiddleWareBuilderExtension

Reviewer: I just checked the the NetTcp and asp.net core equivalents are static classes, and this needs to be static as it contains extension methods.
```suggestion
public static class QueueMiddleWareBuilderExtension
```

{
    internal const string InvokeAsyncMethodName = "InvokeAsync";

>     public static IQueueMiddlewareBuilder UseMiddleware<TMiddleware>(this IQueueMiddlewareBuilder app, params object[] args)

Reviewer: You don't need to stub out the methods or provide internal details. Should this class be static?

    {

>          return app.UseMiddleware(typeof(TMiddleware), args);

Reviewer: Remove implementation detail and leave it as definition only. Just put a semi-colon at the end of the method name like you would an abstract method, this doesn't need to be compilable.

    }
    
    public static IQueueMiddlewareBuilder UseMiddleware(this IQueueMiddlewareBuilder app, Type middleware, params object[] args)
    {
        //TODO
     }
    }
```
#### QueueServiceWebHostBuilderExtension

> - QueueServiceWebHostBuilderExtension is the WebHostBuilder extension which stiches all the middleware pipeline and ```QueuePollingService``` which will be called during service startup for queuing mechanism to work. e:g:- 

Reviewer: Make sure any specified types are quoted as inline code. Eq `QueueServiceWebHostBuilderExtension` as opposed to QueueServiceWebHostBuilderExtension

Reviewer: You should provide the class design here. Also the example should be updated to .NET 6 style syntax. See the sample app [here](https://github.com/CoreWCF/samples/blob/main/Scenarios/Logging-and-Dependency-Injection/LoggingSample/Program.cs) for an example.

Reviewer: We don't need to add CoreWCF into the method name. It's not `UseAspNetCoreKestrel`. We should just be bold and call it `UseQueues`. This is also the wrong place to call the method, as Use verbs are applied to builders, Add verbs to collections. I've been trying to work out where we should apply the `UseQueues` method and I'm beginning to think for CoreWCF 2.0 we need to take a bit of a step back to look at the api's and how you configure CoreWCF in the future. As of right now we have NetTcp (which uses UseConnectionHandler and is based on Kestrel) and HTTP based transports, so everything builds off of a WebHost. We're about to add queue's, and then we have named pipe's coming up, neither of which will build on top of Kestrel. Then there's TCP Port Sharing, and the IIS hosted model in the future where the IIS Integration IServer type is used, both of which will have us using sockets outside of an IServer. There's also been some demand for UDP based bindings too. There's a new model in .NET 6 where a Web based application (mean HTTP based) is just another IHostedService implementation. I think we need to start thinking about configuring CoreWCF transports outside of the IWebHostBuilder world (as you might not have an HTTP endpoint) and have some mechanism build around `IHostBuilder`. I don't want to hold up merging this based on that (as it will take some designing), this is more of a heads up that this portion is likely to evolve. It shouldn't affect the rest of the design as IHostedService is still the right way to go. Basically everything should become an IHostedService and this is around how do we add/configure the various transports in the host. We also need to solve the problem of not wanting to configure your services and endpoints in multiple places.

Author: Agree, this will evolve as we build more and more layers..

```csharp
internal sealed class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
      services.AddServiceModelServices();
      services.UseCoreWCFQueue();
 }
 ```

>  - It first keeps an instance of IQueueMiddlewareBuilder for chaining all the middleware components to be used

Reviewer: This isn't making a lot of sense. Do you mean it adds an implementation of IQueueMiddlewareBuilder to DI so that a Queue based binding is able to build the transport layer stack by adding middleware? It will add other needed components to DI such as QueueHandShakeMiddleWare and QueuePollingService which are necessary to build and run a Queue based service.  

We also need to rename QueueHandShakeMiddle. The NetTcp one is called that because there is a connection handshake which is completed in stages (here's my endpoint, here's my framing mode, here's the stream upgrade mime type I wish to upgrade etc). There is no handshake with Queue. One name that seems appropriately short but descriptive is (and I'm not the only person who can have an opinion on this, so feel free to suggest other names) QueueDispatchMiddleware? QueuePollingService seems like a good name, although QueuePollingHostedService might be a better fit as it describes what kind of service it is (it implements IHostedService).

Author: QueueFetchAndDispatchMiddleware ? => As it describes the fact that it first fetches the queue by using the transport and dispatch service dispatcher to process.
QueuePollingHostedService would be better as it clearly speaks about the IHostedService

 - Afterwards it initalizes the ```QueueHandShakeMiddleWare``` and ```QueuePollingService```.
#### QueueHandShakeMiddleWare

> - The QueueHandShakeMiddleWare has 2 methods, ```InitQueueMiddleWare``` and ```Build```.

Reviewer: We need to provide the class description as code. The name `InitQueueMiddleware` is very redundant. Also as it's called from the constructor, it's an implementation detail and should be a private method. There's no reason for user code to call it.  
Please specify the type of delegate which is returned.

- The InitQueueMiddleWare initalizes middlewares to use and is called from constructor
- Build method builds the middleware pipeline and returns delegate.
#### QueueBaseBindingElement , QueueTransportContext and IQueueTransport

> - IQueueTransport holds the signature for method ```ReceiveQueueMessageContextAsync```. 

Reviewer: ```suggestion
- `IQueueTransport` defines a single method, `ReceiveQueueMessageContextAsync`. The QueuePollingService will call this method to fetch the next incoming queue message.

```csharp
   public interface IQueueTransport
    {
        Task<QueueMessageContext> ReceiveQueueMessageContextAsync(CancellationToken cancellationToken);
    }
```

> - QueueBaseBindingElement is an abstract class with ```CreateQueueTransport``` as abstract method. Any concrete QueueTransport part of the CoreWCF, needs to implement and provide a concrete implementation of IQueueTransport.

Reviewer: The second sentence needs to rewording.

```csharp
    public abstract class QueueBaseBindingElement : BindingElement
    {
        protected QueueBaseBindingElement()
        {
        }

        protected QueueBaseBindingElement(QueueBaseBindingElement elementToBeCloned)
            : base(elementToBeCloned)
        {
        }

        public abstract IQueueTransport CreateQueueTransport(BindingContext context);
        public virtual int MaxPendingReceives { get { return 1; } }
    }
 ```
- The concrete implementation of IQueueInternalTransport will be used by CoreWCF to retrieve messages from the queue storage service. 

> - When ReceiveQueueMessageContextAsync method is called, it should return a a concrete implementation of QueueMessageContext object.

Reviewer: Weren't we going to try to have QueueMessageContext be usable for many scenarios as is and the only reason to derive it would be if a particular implementation needs to do something different?

Author: do you want me to add this ?

> - QueueTransportContext is a holder class to store InternlTransport, QueueBaseBindingElement, ServiceDispatcher

Reviewer: I thought we fixed the naming here. InternalTransport is the wrong name, I thought it was the just the IQueueTransport?

#### QueuePollingService

> - QueuePollingService is the center for queue implementation, it starts when the polling messages would start and when it would end. 

Reviewer: For the second half of this sentence, maybe "it's started when the host starts and continues fetching messages until the host is stopped."

- At the initialization of the class, 
```csharp
private async Task Init()
{

>    foreach (Type serviceType in serviceBuilder.Services)

Reviewer: Nit, indentation

    - If exists, create a service dispatcher.
    - Using the ```QueueBaseBindingElement``` , internal transport (transport layer implemented by MSMQ/RabbitMQ/SQS/AMQP etc) is invoked to provide a concrete internal transport
    - Along with internal transport concrete object, service dispatcher, middle ware delegates is created for each service type and stored as list of ```QueueTransportContext``` object.
-  QueuePollingService implements IHostingService, as a result it implements StartAsync/StopAsync methods. 
-  StartAsync is called by WebHost and it performs below task,
```csharp
public async Task StartAsync(CancellationToken cancellationToken)
{
   var tasks = _queueTransportContexts.Select(_queueTransport => StartFetchingMessage(_queueTransport));
   await Task.WhenAll(tasks); // handle cancellation
}
private async Task StartFetchingMessage(QueueTransportContext queueTransport)
{
    List<Task> tasks = new List<Task>();
    for (int i = 0; i < queueTransport.QueueBindingElement.MaxPendingReceives; i++)
    {
        tasks.Add(FetchAndProcessAsync(queueTransport, _cancellationTokenSource.Token));

> 

Reviewer: Formatting, you have extra blank lines in some of these methods

    }
    await Task.WhenAll(tasks);
}

private async Task FetchAndProcessAsync(QueueTransportContext queueTransport, CancellationToken cancellationToken)
{
    while (true || !cancellationToken.IsCancellationRequested)
    {
        QueueMessageContext queueMessageContext = await queueTransport.QueueInternalTransport.ReceiveQueueMessageContextAsync(cancellationToken);
        queueMessageContext.QueueTransport = queueTransport;
        await queueTransport.QueueHandShakeDelegate(queueMessageContext);
        
    }
}
        
```
- For StartAsync
   - First we create a list of tasks```(StartFetchingMessage)``` for each ```QueueTransportContext``` (populated during Init). 

>    -  Each Task uses internal transport concurrency (MaxPendingReceives) to create another list of subtask to fetch message from each internal transport concurrently```( By FetchAndProcessAsync)

Reviewer: You have an unpaired backtick on this line for code formatting.

 - When the ASP.NET Core WebHost starts, it calls IHostedService.StartAsync on the QueuePollingService, which creates list os tasks (StartFetchingMessage) for each transport and run them in a loop.
 -  Each task again create a list of subtasks based on the QueueTransport  (concurrent property) to retrieve message context(QueueMessageContext) (until server stops)
 - The fetched QueueMessageContext is passed to the middleware. The first middleware layer reads the message bytes from a PipeReader and decodes it to a Message object.
 - The Message object is put into QueueMessageContext object and passed to next layer QueueProcessMessage.
 - The QueueProcessMessage, read the message and invokes DispatchAsync on the IServiceChannelDispatcher to call the service based on the binding and message header information.
 - Once the service is processed, QueueMessageContext sends result to concrete transport to notify that message is being processed. It will either be a NullMessage meaning everything was successful, or  it will be a message fault which indicates an error processing the request.

> # Nofiying Client

Reviewer: Did you mean notifying?

        protected QueueBaseTransportBindingElement(QueueBaseTransportBindingElement elementToBeCloned)
            : base(elementToBeCloned)
        {
        }

        public virtual int MaxPendingReceives { get { return 1; } }
        public abstract QueueTransportPump BuildQueueTransportPump(BindingContext context);

        public override bool CanBuildServiceDispatcher<TChannel>(BindingContext context)
        {
            return (typeof(TChannel) == typeof(IInputChannel));
        }

        public override IServiceDispatcher BuildServiceDispatcher<TChannel>(BindingContext context, IServiceDispatcher innerDispatcher)
        {
            return innerDispatcher;
        }

    }
 ```
- The concrete implementation of `IQueueTransport` will be used by CoreWCF to retrieve messages from the queue storage service. 
- When `ReceiveQueueMessageContextAsync` method is called, it returns an instance of `QueueMessageContext` object.
- `ServiceDispatcher`, `QueueBaseTransportBindingElement`, `QueueMessageDispatcherDelegate`, and `QueueTransportPump` are all stored in `QueueTransportContext`. 
-`QueueTransportPump` is an abstract class with `StartPumpAsync` and `StopPumpAsync` task. 

> - Any queue transport which is using push mechanism (e:g:- RabbitMQ) needs to implement this class.

Reviewer: Inconsistent spacing between - and text on this line. Also, StartPumpAsync and StopPumpAsync are methods, not tasks. Maybe refer to them as "Task returning methods". Technically calling them async methods isn't accurate as they could be implemented synchronously, they just need to return a Task.

            _queueTransportContexts.Add(new QueueTransportContext
            {
                QueuePump = queuePump,
                 ServiceDispatcher = _serviceDispatcher,
                 QueueBindingElement = queueTransportBinding,
                 MessageEncoderFactory = msgEncBindingelement.CreateMessageEncoderFactory(),
                 QueueHandShakeDelegate = _queueHandShakeMiddleWare.Build()
            });

        }
    }

}
```
- for each service type (which is fetched from DI), service dispatcher is created.
- for each service dispatcher, validation is made to ensure the service dispatcher binding element collection property has a transport binding element of type `QueueBaseTransportBindingElement`.
- using the `QueueBaseTransportBindingElement` which is fetched from above step, `BuildQueueTransportPump` method is called to get the `QueueTransportPump`.
 - In the end, for every service dispatcher, a `QueueTransportContext` object is created which stores objects of type `QueueTransportPump`, `ServiceDispatcher`, `QueueBaseTransportBindingElement`,`MessageEncoderFactory` and `QueueHandShakeDelegate`.

> - Since `QueuePollingService` implements `IHostingService`, it implements the `StartAsync` and `StopAsync` methods.

Reviewer: Extra space at start of line

- `StartAsync` is called by WebHost and it performs the below task.
```csharp
public async Task StartAsync(CancellationToken cancellationToken)
{
    var tasks = _queueTransportContexts.Select(_queueTransport => StartFetchingMessage(_queueTransport));
    await Task.WhenAll(tasks);
}

private async Task StartFetchingMessage(QueueTransportContext queueTransport)
{
    await queueTransport.QueuePump.StartPumpAsync(queueTransport, CancellationToken.None);
}

```
- For StartAsync
 - First we create a list of tasks`StartFetchingMessage` for each `QueueTransportContext` (populated during Init).
 - `QueueTransportPump` from `QueueTransportContext` is used by each `StartFetchingMessage` to call the `StartPumpAsync` method.
 - If the `QueueTransportPump` is based on a push message mechanism, it awaits until a message is received and invokes service as soon as the message is received.

>  - If the `QueueTransportPump` is using `DefaultQueueTransportPump`, it creates multiple concurrent tasks to invoke `FetchAndProcessAsync` method based on concurrency level.

Reviewer: If the `QueueTransportPump` is based on a push message mechanism, the provided implementation awaits...  

For push, the implementer needs to provide a concrete implementation instead of proving an in-box implementation. Without this small change in wording, I think it might come across as there's one in box which does this already.

Reviewer: I also think saying "awaits" implies an implementation which might not be the case. With the EventHandler mechanism some libraries use you aren't awaiting anything. Maybe replace awaits with waits as that's more generic, and awaiting is a type of waiting, as is an EventHandler being executed. I know that's a lot of words to say awaits -> waits, but wanted to give my rationale.

 -The `FetchAndProcessAsync` method of QueueTransport (concrete transport) calls `ReceiveQueueMessageContextAsync` to retrieve the message and pass the handle to the middleware for processing. 
- StopAsync cancels all tasks and, as a result, the fetching of messages is stopped.
#### QueueFetchMessage
- `QueueFetchMessage` is the first middleware which is invoked as soon as a message is fetched from the queue.
- The `InvokeAsync` method reads messages from pipereader and encodes them into a concrete `Message` objet and puts them back to `QueueMessageContext` and invokes the next delegate , which is `QueueProcessMessage`.
#### QueueProcessMessage
- Invokes the `DispatchAsync` method of the channel dispatcher by passing `QueueMessageContext` object and ending the middleware pipeline.

> - As a part of `DispatchAsync`, the service operation is called, processing of messages is done, eventually `QueueMessageContext` (inherits `RequestContext`) `ReplyAsync` method is called.

Reviewer: ending the queue middleware pipeline and transitions to the dispatch pipeline

- The `ReplyAsync` mechanism is used to inform the concrete transport that a message processing has finished with a status of success or failure by invoking an Action `DispatchResultHandler`. 

> - By calling the `DispatchResultHandler` each queue providers which support timeouts or requires notification to prevent reprocessing a message can take appropriate actions. 

Reviewer: inform the transport implementation that message processing has finished with a status of success of failure. It signals the status result by invoking a `DispatchResultHandler` delegate.

> - The reply message sent via `DispatchResultHandler` could be a fault message, which will signal a failure to process the message, and can be used to provide an error message back to the queue provider.

Reviewer: Flip the sentence around:  
If a queue provider requires completion notification to support timeouts or other features, they can do so by setting the `DispatchResultHandler` delegate and taking action accordingly.

# Workflow Diagram
![CoreWCF Generic Queue](/Documentation/DesignDocs/corewcf_queue.png?raw=true)

 - During the server registrations, the QueuePollingService initializes the service dispatcher and gets the concrete transport from each queue layer transport.
 - When the ASP.NET Core WebHost starts, it calls IHostedService. StartAsync on the QueuePollingService, creates a list of tasks (StartFetchingMessage) for each transport and runs them in a loop.

>  - Each task again creates a list of subtasks based on the QueueTransport (concurrent property) to retrieve message context(QueueMessageContext) (until server stops)

Reviewer: Remove space between IHostedService and StartAsync, and backtick quote as code. Same with QueuePollingService and StartFetchingMessage

>  - The fetched QueueMessageContext is passed to the middleware. The first middleware layer reads the message bytes from a PipeReader and decodes them to a Message object.

Reviewer: Name the concurrent property with ClassName.PropertyName and backtick quote.. Remove (until server stops) from parenthesis and instead add a comma just before that phrase. Also worth mentioning until Webhost calls QueuePollingService.StopAsync as the mechanism that it will be stopped.

>  - The Message object is put into the QueueMessageContext object and passed to the next layer, QueueProcessMessage.

Reviewer: There are multiple types of middleware so worth disambiguating by calling it the queue middleware or queuing middleware (make sure it's consistent through document.). Might be worth specifying earlier in the document where you are talking about building this middleware that this middleware pipeline will be referred to as the queue middleware.

 - The QueueProcessMessage reads the message and invokes DispatchAsync on the IServiceChannelDispatcher to call the service based on the binding and message header information.

>  - Once the service is processed, QueueMessageContext sends the result to concrete transport to notify that the message is being processed. It will pass status (failed/success) and a full message , so that concrete queue transport can take appropriate actions.

Reviewer: Reads the message using the binding provided MessageEncoder.

> 

Reviewer: Once the service request, or maybe just request. "Concrete" typically makes more sense when talking about type hierarchies. Here it sounds better to say the queue transport implementation.  
It will pass a result status (failed/success) and an optional fault Message so that the queue transport implementation... (remove comma)

# Notifying Client
- Every internal transport when constructing a `QueueMessageContext` via `ReceiveQueueMessageContextAsync` method, they can pass a delegate in the form of `DispatchResultHandler` .
- Once a message is processed and service is invoked, the control comes to `ReplyAsync` method of `QueueMessageContext`, where if `DispatchResultHandler` is defined, it's invoked by passing a) status of service invocation and b)full `QueueMessageContext`

> - Internal transport can act based on that to notify client or put into dead letter queue or trigger any custom fuctions.

Reviewer: ...method of `QueueMessageContext`. If `DispatchResultHandler` has been set, it's invoked...  
Defined would mean whether the property or field exists on the class, not whether it holds a value. Also don't use a) and b) to list passed parameters, just list them with commas.
